File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    answer = []
    visited = [False] * len(tickets)
    
    def dfs(airport, path):
        # 모든 티켓을 한번씩 다 사용해서 갈 수 있는 모든 경로가 answer에 저장됨
        if (len(path) == len(tickets) + 1):
            answer.append(path)
            return
        for idx, ticket in enumerate(tickets):
            if ticket[0] == airport and visited[idx] == False:
                visited[idx] = True
                dfs(ticket[1], path + [ticket[1]])
                visited[idx] = False
    dfs("ICN", ["ICN"])
    # 저장 된 모든 경로 중에 알파벳 순서가 앞서는 순서들로 이뤄진 첫번째 경로 반환하기 위해서
    answer.sort()
    return answer[0]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    boolean[] visited;
    ArrayList<String> route;
    
    public String[] solution(String[][] tickets) {
        String[] answer = {};
        int cnt = 0;
        visited = new boolean[tickets.length];
        route = new ArrayList<>();
        
        dfs("ICN", "ICN", tickets, cnt);
        
        Collections.sort(route);
        
        answer = route.get(0).split(" ");
        
        return answer;
    }
    
    public void dfs(String start, String path, String[][] tickets, int cnt) {
        if (cnt == tickets.length) {
            route.add(path);
            return;
        }
        
        for (int i = 0; i < tickets.length; i++) {
            if (start.equals(tickets[i][0]) && !visited[i]) {
                visited[i] = true;
                dfs(tickets[i][1], path + " " + tickets[i][1], tickets, cnt+1);
                visited[i] = false;
            }
        }
    }
}